controllers/integration/integration_adapter.go

```diff
@@ -137,6 +137,139 @@ func (a *Adapter) getAllApplicationSnapshots() (*[]appstudioshared.ApplicationSn
 	return &applicationSnapshots.Items, nil
 }
 
+func (a *Adapter) getAllEnvironments() (*[]appstudioshared.Environment, error) {
+
+	environmentList := &appstudioshared.EnvironmentList{}
+	opts := []client.ListOption{
+		client.InNamespace(a.application.Namespace),
+	}
+	err := a.client.List(a.context, environmentList, opts...)
+	return &environmentList.Items, err
+}
+
+func (a *Adapter) findAvailabeEnvironments() (*[]appstudioshared.Environment, error) {
+	allEnvironments, err := a.getAllEnvironments()
+	if err != nil {
+		return nil, err
+	}
+	availableEnvironments := []appstudioshared.Environment{}
+	for _, environment := range *allEnvironments {
+		if environment.Spec.ParentEnvironment == "" {
+			availableEnvironments = append(availableEnvironments, environment)
+		} else {
+			for _, tag := range environment.Spec.Tags {
+				if tag == "ephemeral" {
+					availableEnvironments = append(availableEnvironments, environment)
+				}
+			}
+		}
+	}
+	return &availableEnvironments, nil
+}
+
+func (a *Adapter) findExistingApplicationSnapshotEnvironmentBinding(environmentName string) (*appstudioshared.ApplicationSnapshotEnvironmentBinding, error) {
+	applicationSnapshotEnvironmentBindingList := &appstudioshared.ApplicationSnapshotEnvironmentBindingList{}
+	opts := []client.ListOption{
+		client.InNamespace(a.application.Namespace),
+		client.MatchingFields{"spec.application": a.application.Name},
+		client.MatchingFields{"spec.environment": environmentName},
+	}
+
+	err := a.client.List(a.context, applicationSnapshotEnvironmentBindingList, opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	return &applicationSnapshotEnvironmentBindingList.Items[0], nil
+}
+
+func (a *Adapter) EnsureApplicationSnapshotEnvironmentBindingExist() (results.OperationResult, error) {
+	availableEnvironments, err := a.findAvailabeEnvironments()
+	if err != nil {
+		return results.RequeueWithError(err)
+	}
+
+	applicationSnapshot, err := a.findMatchingApplicationSnapshot()
+	if err != nil {
+		return results.RequeueWithError(err)
+	}
+
+	components, err := a.getAllApplicationComponents(a.application)
+
+	if err != nil {
+		return results.RequeueWithError(err)
+	}
+	for _, availableEnvironment := range *availableEnvironments {
+		bindingName := a.application.Name + "-" + availableEnvironment.Name + "-" + "binding"
+		applicationSnapshotEnvironmentBinding, _ := a.findExistingApplicationSnapshotEnvironmentBinding(availableEnvironment.Spec.DisplayName)
```

Comment:
Looks like we'll also need to change the `availableEnvironment.Spec.DisplayName` to `availableEnvironment.Name` here.

Reply:
@dirgim `availableEnvironment.Name` is unique and created by system, right?

Reply:
Yes, that's correct, the `Name` of the resource must be unique

Reply:
Thank you.

Reply:
Updated

Reply:
Hi @jinqi7 I'm still seeing `availableEnvironment.Spec.DisplayName` here, can you check if it's the same on your side?

Reply:
OK. Updated

Reply:
Thanks, looks good now